### PR TITLE
Graph comparing models over latent dimensions 

### PR DESCRIPTION
**Pull request by jaceybronte**
This PR optimizes each model for a certain number of latent dimensions and then graphs to compare.

### CODE REVIEW

File: 2.train-VAE/utils/betatcvae.py [hunk 1]
```diff
@@ -334,9 +334,10 @@ def tc_weights(model, subset_train_df, path=None):
     # Create a dataframe from the weight matrix
     weight_df = pd.DataFrame(weight_matrix)
 
-    # Save as parquet to use for heatmap
-    weight_df_dir = pathlib.Path("./results/tc_weight_matrix_encoder.parquet")
-    weight_df.to_parquet(weight_df_dir)
+    if path:
```
Comment (gwaybio) [inline 2.train-VAE/utils/betatcvae.py:337]: Maybe something to consider when/if the times comes around for a refactor, but a lot of these changes are the same, except the name of the parquet file

i haven't looked deeply into the contents of the utils folder, but I imagine that much of these model-specific utils can be abstracted into a more versatile class. This way, any change you make to one utils, it will be automatically propagated to others (so you don't have to manually maintain consistency)

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Can you add a bit more details to a markdown cell above describing what this notebook accomplishes? I assume that it is generating the latent dimension dataframe that you're using to visualize? My assumptions can be confirmed also by renaming the notebook file - I often find that verbs are good for notebook names! (e.g., `0.generate_latent_dimension_dataframe,ipynb`)

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+output_dir = pathlib.Path("./results").resolve()
+output_dir.mkdir(parents=True, exist_ok=True)
+
+
+# In[6]:
+
+
+# Function to perform dimensionality reduction and extract z matrices for sklearn models
+def extract_z_matrix_sklearn(model, data, model_name, num_components, output_dir):
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: neat function! However, it should be defined at the beginning on the notebook (after imports)

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    if model_name == "nmf":
+        min_value = data.min().min()
+        if min_value < 0:
+            data = data - min_value
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: I'm a bit confused why this specific step is necessary? I understand NMF needs to have all positive values, but in the past we've performed a 0-1 normalization (see https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) (this is in contrast with the z-score norm we perform for other models)

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    
+    return z_matrix_df
+
+
+# In[7]:
+
+
+def extract_weights(model, model_name, H=None):
+    if model_name in ["pca", "ica", "nmf"]:
+        # Transpose PCA components and format columns
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Nice commenting!

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        weights_df.columns = [f"{x}" for x in range(0, weights_df.shape[1])]
+        
+    # Reset index to rename 'index' to 'genes'
+    weights_df = weights_df.reset_index().rename(columns={"index": "genes"})
+    
+    return weights_df
+
+
+def perform_gsea(weights_df, model_name, num_components):
+    library = blitz.enrichr.get_library("Reactome_2022")
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: I recommend using this library as a function argument (can set as default, but it will make for more versatility)

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    seed = random.random()
+    gsea_results = []
+    # Define cutoff values
+    lfc_cutoff = 0.584
+    fdr_cutoff = 0.25
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: How did you define these? Also, I recommend making these function arguments

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+                        "z": int(col),
+                        "full_model_z": num_components,
+                        "model": str(model_name),
+                        "reactome pathway": str(pathway_result['Term']),
+                        "gsea es score": pathway_result['es'],
+                        "nes score": pathway_result['nes'],
+                        "p value": pathway_result['pval'],
+                        "shuffled": False
+                    }
+                    gsea_results.append(result_row)
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Not sure if you need to do all of this? Is `gsea_result` already a pandas dataframe? Is this loop renaming columns and adding new ones?

Can you do something like this then?

```python
gsea_result = (
    gsea_result.rename(
        columns={
            "model": "model_rename"
        }
    )
    .assign(shuffled=False)
)
```

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+                    
+            except ZeroDivisionError:
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: nice use of exceptions!

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        z_matrix_df = None
+        
+        if model_name in ["pca", "ica", "nmf"]:
+            # Sklearn models (PCA, ICA, NMF)
+            if model_name == "pca":
+                model = PCA(n_components=num_components)
+            elif model_name == "ica":
+                model = FastICA(n_components=num_components)
+            elif model_name == "nmf":
+                model = NMF(n_components=num_components, init='nndsvd', max_iter=1000, random_state=0)
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Any specific reason to use this `init` method? (just curious!)

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+            if ((combined_results_df['model'] == model_name) & 
+                (combined_results_df['full_model_z'] == num_components)).any():
+                print(f"Skipping {model_name} with {num_components} dimensions as it is already processed.")
+                continue  # Skip to the next iteration if this combination is already present
+        z_matrix_df = None
+        
+        if model_name in ["pca", "ica", "nmf"]:
+            # Sklearn models (PCA, ICA, NMF)
+            if model_name == "pca":
+                model = PCA(n_components=num_components)
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Oh, are you performing the model training here too? This seems like a very complicated notebook! (perhaps my name suggestion above is not accurate)

In general, my recommended workflow is to train separately. You can save the models (as joblib files, for example) and then load them in later, when you're performing eval. (it's also possible my read on this section of the notebook is incorrect, so please correct me if necessary)

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+
+#Save as CSV for R 
+csv_output_file = output_dir / "combined_z_matrix_gsea_results.csv"
+combined_results_df.to_csv(csv_output_file, index=False)
+
+
+# In[9]:
+
+
+combined_results_df.sort_values(by='gsea es score', key=abs, ascending = False).head(50)
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: I must have missed this earlier in the notebook, but try to avoid using spaces in column names (make it `gsea_es_score`). Spaces remove functionality in pandas (specifically, column referencing with dot notation [e.g., `combined_results_df.gsea es score` does not work])

File: 4.gene_expression_signatures/scripts/1.latent_dim_plot.r [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    results_file,
+)
+
+
+# In[3]:
+
+
+# Prepare the data
+# Filter for the single highest ES score for each model at each dimension
+max_es_df <- gsea_results_df %>%
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/1.latent_dim_plot.r]: nice use of `dplyr`!!!

File: 4.gene_expression_signatures/scripts/1.latent_dim_plot.r [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+                      "#e4e716", 
+                      "#984ea3", 
+                      "#ff7f00"),
+                     labels = c(
+                      "pca" = "PCA", 
+                      "ica" = "ICA", 
+                      "nmf" = "NMF", 
+                      "vanillavae" = "VAE",
+                      "betavae" = "BVAE", 
+                      "betatcvae" = "BTCVAE"
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/1.latent_dim_plot.r]: I recommend saving these sorts of variables in a `themes.R` utility file, so that you can reference them in all future notebook calls. E.g., See https://github.com/WayScience/phenotypic_profiling/blob/main/7.figures/themes.r

File: 4.gene_expression_signatures/scripts/1.latent_dim_plot.r [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+                      "betatcvae" = "BTCVAE")) +
+  labs(x = "Latent Dimensions", y = "Highest ES Score", title = "Highest ES Score by Latent Dimension for Each Model") +
+  theme(legend.position = "right")
+
+
+# In[5]:
+
+
+# Save the plot with custom dimensions
+ggsave("latent_plot.png", plot = latent_plot, width = 10, height = 8, units = "in")
```
Comment (gwaybio) [inline 4.gene_expression_signatures/scripts/1.latent_dim_plot.r]: These look so so cool!!!!!! Very exciting to see this.

Upon closer inspection, however, I see the purple PCA get a little wonky at latent dim = 100. Do you know what is going on there?

**Review by gwaybio — CHANGES_REQUESTED** [review]
This is a really nice pull request that was a pleasure to review. Well done! I made several comments, many which should be addressed prior to merging. Others feel free to add a discussion to them, if you feel they are out of scope for this PR.

File: 2.train-VAE/utils/betatcvae.py (hunk 1 above)
Comment (jaceybronte) [reply 2.train-VAE/utils/betatcvae.py:337]: I agree yes, there is definitely some redundancy that could be improved upon and I hope to do a PR in the future with that goal, but I think is probably out of scope for this one and would need to be its own PR!

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 3 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Fixed!

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 6 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Done

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 7 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: I ended up just removing the significance portion as its really not necessary, as the significant dataframe isn't returned

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 10 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: At a certain number of components the NMF model was running into sparseness errors, which is best for the nndsvd initialization rather than random according to the documentation

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 12 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Fixed!

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 11 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Separated this!

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 2 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Separated out and renamed!

File: 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py (hunk 4 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py]: Yep the data was already normalized with zero_one_normalize in the load data section, not sure why I had this but it's been removed!

File: 4.gene_expression_signatures/scripts/1.latent_dim_plot.r (hunk 15 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/1.latent_dim_plot.r]: I reran some things, and there is still a PCA outlier at z=45 - not sure what it is about!

File: 4.gene_expression_signatures/scripts/1.latent_dim_plot.r (hunk 14 above)
Comment (jaceybronte) [reply 4.gene_expression_signatures/scripts/1.latent_dim_plot.r]: Done!

**Comment by jaceybronte** [comment]
@gwaybio ready for rereview!

File: utils/themes.r [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+  "betavae" = "BVAE", 
+  "betatcvae" = "BTCVAE"
+)
+
+# Custom theme function
+custom_theme <- function() {
+  theme(
+    legend.position = "right"
+  )
+}
```
Comment (gwaybio) [inline utils/themes.r]:
```suggestion
}

```

to remove nasty red symbol, always need an empty new line at the end of scripts

**Review by gwaybio — APPROVED** [review]
LGTM 🎉
